### DISCUSSION TITLE
How to run same telegram bot on multiple app instances ?

### DISCUSSION THREAD

**Discussion by d-kobzar** [nksmnf/nestjs-telegraf#1147] [Q&A]
In my case i need to run e.g. `notification-service` to send messages to the users in telegram bot. But also i need to handle their actions in `api-service` which have controllers. But in case i try to start my apps, then i receive error in console `TelegramError: 409: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running`.

I've found that it's possible to run 1st instance with launching bot and 2nd instance of app without launching. This case is described here https://stackoverflow.com/a/71466482 . But in nestjs-telegraf it's impossible to run module without launching.
How can i solve this issue ?

**Comment by backrebital-lgtm** [comment]
`409 Conflict` is expected when two processes call `getUpdates`: Telegram permits only one long-polling consumer per bot token. You can still run the Nest app on every replica; start polling on exactly one replica (or on the current leader) and let the others send messages only.

`nestjs-telegraf` has an explicit switch for this:

```ts
// Set TELEGRAM_POLLING=true only on the elected leader.
TelegrafModule.forRoot({
  token: process.env.BOT_TOKEN!,
  launchOptions:
    process.env.TELEGRAM_POLLING === 'true'
      ? { dropPendingUpdates: false }
      : false,
});
```

`launchOptions: false` skips `bot.launch()` while still creating the bot and wiring the Nest listeners. Therefore `@InjectBot()` can be used on every replica for `sendMessage`, `editMessageText`, etc.; only the polling replica should consume updates. Do not set the flag to `true` on every pod.

For failover, make the flag come from real leader election/lease ownership (for example a Kubernetes lease), not a best-effort Redis lock, and ensure only one holder is active at a time. The process needs to restart/reconfigure when leadership changes. If you need several replicas to process incoming updates concurrently, use webhook mode behind a load balancer plus a durable queue/idempotency key; also keep session state in Redis/DB rather than process memory. Telegram retries webhook deliveries, so handlers should be idempotent.
